File: main/csharp/ir/HeaderStructure.cs
using System.Collections.Generic;
using Adaptive.SimpleBinaryEncoding.Util;

namespace Adaptive.SimpleBinaryEncoding.ir
{
    /// <summary>
    /// Metadata description for a message headerStructure
    /// </summary>
    public class HeaderStructure
    {
        internal const string BlockLength = "blockLength";
        internal const string TemplateId = "templateId";
        internal const string SchemaId = "schemaId";
        internal const string SchemaVersion = "version";

        private readonly IList<Token> _tokens;
        private PrimitiveType _blockLengthType;
        private PrimitiveType _templateIdType;
        private PrimitiveType _schemaIdType;
        private PrimitiveType _schemaVersionType;

        internal HeaderStructure(IList<Token> tokens)
        {
            Verify.NotNull(tokens, "tokens");
            _tokens = tokens;

            CaptureEncodings(tokens);

            Verify.NotNull(_blockLengthType, "blockLengthType");
            Verify.NotNull(_templateIdType, "templateIdType");
            Verify.NotNull(_schemaIdType, "schemaIdType");
            Verify.NotNull(_schemaVersionType, "schemaVersionType");
        }

        private void CaptureEncodings(IList<Token> tokens)
        {
            foreach (Token token in tokens)
            {
                switch (token.Name)
                {
                    case BlockLength:
                        _blockLengthType = token.Encoding.PrimitiveType;
                        break;

                    case TemplateId:
                        _templateIdType = token.Encoding.PrimitiveType;
                        break;

                    case SchemaId:
                        _schemaIdType = token.Encoding.PrimitiveType;
                        break;

                    case SchemaVersion:
                        _schemaVersionType = token.Encoding.PrimitiveType;
                        break;
                }
            }
        }

        /// <summary>
        /// List of <see cref="Token"/> associated with this <see cref="HeaderStructure"/>
        /// </summary>
        public IList<Token> Tokens
        {
            get { return _tokens; }
        }

        /// <summary>
        /// Underyling type used to store block length information
        /// </summary>
        public PrimitiveType BlockLengthType
        {
            get { return _blockLengthType; }
        }

        /// <summary>
        /// Underyling type used to store template id infomation
        /// </summary>
        public PrimitiveType TemplateIdType
        {
            get { return _templateIdType; }
        }

        /// <summary>
        /// Underyling type used to store schema id information
        /// </summary>
        public PrimitiveType SchemaIdType
        {
            get { return _schemaIdType; }
        }

        /// <summary>
        /// Underyling type used to store schema version information
        /// </summary>
        public PrimitiveType SchemaVersionType
        {
            get { return _schemaVersionType; }
        }
    }
}

File: main/java/uk/co/real_logic/sbe/ir/HeaderStructure.java
/*
 * Copyright 2013 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.ir;

import uk.co.real_logic.sbe.PrimitiveType;
import uk.co.real_logic.sbe.util.Verify;

import java.util.List;

/**
 * Metadata description for a message headerStructure
 */
public class HeaderStructure
{
    public static final String BLOCK_LENGTH = "blockLength";
    public static final String TEMPLATE_ID = "templateId";
    public static final String SCHEMA_ID = "schemaId";
    public static final String SCHEMA_VERSION = "version";

    private final List<Token> tokens;
    private PrimitiveType blockLengthType;
    private PrimitiveType templateIdType;
    private PrimitiveType schemaIdType;
    private PrimitiveType schemaVersionType;

    public HeaderStructure(final List<Token> tokens)
    {
        Verify.notNull(tokens, "tokens");
        this.tokens = tokens;

        captureEncodings(tokens);

        Verify.notNull(blockLengthType, "blockLengthType");
        Verify.notNull(templateIdType, "templateIdType");
        Verify.notNull(schemaIdType, "schemaIdType");
        Verify.notNull(schemaVersionType, "schemaVersionType");
    }

    private void captureEncodings(final List<Token> tokens)
    {
        for (final Token token : tokens)
        {
            switch (token.name())
            {
                case BLOCK_LENGTH:
                    blockLengthType = token.encoding().primitiveType();
                    break;

                case TEMPLATE_ID:
                    templateIdType = token.encoding().primitiveType();
                    break;

                case SCHEMA_ID:
                    schemaIdType = token.encoding().primitiveType();
                    break;

                case SCHEMA_VERSION:
                    schemaVersionType = token.encoding().primitiveType();
                    break;
            }
        }
    }

    public List<Token> tokens()
    {
        return tokens;
    }

    public PrimitiveType blockLengthType()
    {
        return blockLengthType;
    }

    public PrimitiveType templateIdType()
    {
        return templateIdType;
    }

    public PrimitiveType schemaIdType()
    {
        return schemaIdType;
    }

    public PrimitiveType schemaVersionType()
    {
        return schemaVersionType;
    }
}
